apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/json/MapJsonReader.kt

```diff
@@ -311,25 +357,30 @@ class MapJsonReader(val root: Map<String, Any?>) : JsonReader {
    * Rewinds to the beginning of the current object.
    */
   override fun rewind() {
-    indexStack[stackSize - 1] = 0
+    val container = containerStack[stackSize - 1]
+    path[stackSize - 1] = null
+    iteratorStack[stackSize - 1] = container!!.iterator()
     nameIndexStack[stackSize - 1] = 0
-    rewindPath()
+    advanceIterator()
   }
 
-  private val path = mutableListOf<Any>()
-
   override fun getPath(): String {
-    // Drop the first "."
-    return path.drop(1).joinToString("")
-  }
-
-  private fun incrementPathIndex() {
-    if (path.lastOrNull() is Int) path[path.lastIndex] = (path.last() as Int) + 1
-  }
-
-  private fun rewindPath() {
-    while (path.isNotEmpty() && path.last() != ".") path.removeLast()
-    path.removeLastOrNull()
+    return buildString {
+      for (index in 0.until(stackSize)) {
+        val element = path[index]
+        if (element is String) {
+          if (index > 0) {
+            append('.')
+          }
+          append(element)
+        } else if (element is Int) {
+          append("[$element]")
```

Comment:
```suggestion
          append(".$element")
```

To be the same format as `BufferedSourceJsonReader`

Reply:
Also it looks like there's a +1 in there somewhere, with `PathTest` we can see the indices are 1 too high 😓

Reply:
Aaarg, I've run the wrong test... Ok let me double check that.

Reply:
I've also just realized that the current `getPath()` as a String is pretty bad for `@defer` because these 2 things cannot be distinguished:

```json
{
  "0": true
}
```

and 

```
[
  true
]

```

Reply:
But "0" is not a valid GraphQL field name so it's ok, right?

Reply:
I guess so...